src/v/kafka/server/fetch_memory_units.cc

```diff
@@ -0,0 +1,176 @@
+  ssx::semaphore& kafka_units,
+  ssx::semaphore& fetch_units,
+  local_instance_fn&& local_fn)
+  : _kafka_units(kafka_units)
+  , _fetch_units(fetch_units)
+  , _release_units_timer([this] { release_units(); })
+  , _local_instance_fn(std::move(local_fn)) {
+    _release_units_timer.arm(max_release_period);
+}
+
+ss::future<> fetch_memory_units_manager::stop() {
+    _release_units_timer.cancel();
+    _units_to_release.clear();
+    return ss::now();
+}
+
+void fetch_memory_units_manager::units::adopt(
+  fetch_memory_units_manager::units&& o) {
+    // Adopts assert internally that the units are from the same semaphore.
+    // So there is no need to assert that they are from the same shard here.
+    kafka_units.adopt(std::move(o.kafka_units));
+    fetch_units.adopt(std::move(o.fetch_units));
+}
+
+void fetch_memory_units_manager::release_units() {
+    _release_units_timer.cancel();
+    _release_units_timer.arm(max_release_period);
+
+    _units_to_release.clear();
+}
+
+fetch_memory_units_manager::units&
+fetch_memory_units_manager::units::operator=(units&& o) noexcept {
+    if (this != &o) {
+        this->~units();
+        new (this) units(std::move(o));
+    }
+    return *this;
+}
+
+fetch_memory_units_manager::units::~units() noexcept {
+    if (shard == ss::this_shard_id() || !has_units()) {
+        return;
+    }
+
+    auto f = ss::smp::submit_to(
+      shard,
+      [uk = std::move(kafka_units),
+       uf = std::move(fetch_units)]() mutable noexcept {
+          uk.return_all();
+          uf.return_all();
+      });
+
+    if (!f.available()) {
+        ss::engine().run_in_background(std::move(f));
+    }
+}
+
+fetch_memory_units fetch_memory_units_manager::allocate_memory_units(
+  size_t max_units, const size_t min_units, const bool require_min_units) {
+    const size_t available_units = std::min(
+      _kafka_units.current(), _fetch_units.current());
+    // Note that it's not currently enforced that max_units >= min_units. Hence
+    // we set max_units to the larger of the two here to ensure that is the
+    // case.
+    max_units = std::max(max_units, min_units);
+
+    size_t units_to_alloc = 0;
+    if (require_min_units) {
+        // if \ref require_min_units is true then we must read at least \ref
+        // min_units. So allocate at least that many even if it causes the
+        // semaphores to become negative.
+        units_to_alloc = std::max(
+          min_units, std::min(max_units, available_units));
+    } else if (available_units >= min_units) {
+        // only reserve memory if we have space for at least \ref min_units,
+        // otherwise allocate none.
+        units_to_alloc = std::min(available_units, max_units);
+    }
+
+    if (units_to_alloc == 0) {
+        return {};
+    }
+
+    return {allocate_units(units_to_alloc), _local_instance_fn};
+}
+
+void fetch_memory_units_manager::release_units(units&& u) {
+    auto map_it = _units_to_release.find(u.shard);
```

Comment:
This can be simplified and made more efficient (I think) with try_emplace. You try_emplace the (shard, args for default constructor) and then unconditionally call adopt on the returned iterator.